spec/models/job_preferences/job_scope_spec.rb

```diff
@@ -0,0 +1,205 @@
+    end
+
+    %w[head_of_year head_of_department].each do |granular_role|
+      context "with a jobseeker '#{granular_role}' role split from 'middle_leader' role" do
+        let(:jobseeker_roles) { ["teacher", granular_role] }
+
+        it "returns vacancies with 'middle_leader' role" do
+          expect(job_scope.call).to contain_exactly(vacancy1, vacancy4)
+        end
+      end
+    end
+
+    %w[headteacher headteacher_deputy headteacher_assistant].each do |granular_role|
+      context "with a jobseeker '#{granular_role}' role split from 'senior_leader' role" do
+        let(:jobseeker_roles) { ["teacher", granular_role] }
+
+        it "returns vacancies with 'senior_leader' role" do
+          expect(job_scope.call).to contain_exactly(vacancy1, vacancy3)
+        end
+      end
+    end
+
+    context "with jobseeker roles that don't match any vacancy roles" do
+      let(:jobseeker_roles) { %w[sendco] }
+
+      it "returns no vacancies" do
+        expect(job_scope.call).to be_empty
+      end
+    end
+
+    context "with jobseeker without roles" do
+      let(:jobseeker_roles) { [] }
+
+      it "returns no vacancies" do
+        expect(job_scope.call).to be_empty
+      end
+    end
+  end
+
+  describe "phases scope" do
+    before { job_preferences.phases = jobseeker_phases }
+
+    context "with all jobseeker phases matching vacancy phases" do
+      let(:jobseeker_phases) { %w[primary secondary] }
+
+      it "returns any vacancies with a phase matching the jobseeker phases" do
+        expect(job_scope.call).to contain_exactly(vacancy1, vacancy2, vacancy3, vacancy4)
+      end
+    end
+
+    context "with some of the jobseeker phases matching vacancy phases" do
+      let(:jobseeker_phases) { %w[primary middle] }
+
+      it "returns any vacancies with all their phases included by the jobseeker phases" do
+        expect(job_scope.call).to contain_exactly(vacancy3)
+      end
+    end
+
+    context "with no jobseeker phases matching the vacancy phases" do
+      let(:jobseeker_phases) { %w[middle] }
+
+      it "returns no vacancies" do
+        expect(job_scope.call).to be_empty
+      end
+    end
+  end
+
+  describe "key stages scope" do
+    before { job_preferences.key_stages = jobseeker_key_stages }
+
+    context "with all jobseeker key stages matching vacancy key stages" do
+      let(:jobseeker_key_stages) { %w[ks1 ks2] }
+
+      it "returns any vacancies without any key stages or with all their key stages included in the jobseeker key stages" do
+        expect(job_scope.call).to contain_exactly(vacancy1, vacancy2, vacancy3, vacancy4)
+      end
+    end
+
+    context "with some of the jobseeker key stages matching vacancy key stages" do
+      let(:jobseeker_key_stages) { %w[ks1 ks3] }
+
+      it "returns any vacancies without any key stages or with all their key stages included in the jobseeker key stages" do
+        expect(job_scope.call).to contain_exactly(vacancy1, vacancy2)
```

Comment:
Like the above comment, not related to this change but I would expect `vacancy4` to show here. I may well be misunderstanding though!

Reply:
As mentioned above, is due to the query in `JobScope` using `<@` operator. `vacancy4` has `ks2` key_stage and is missing from the `ks1, ks3` jobseeker ones. So it doesn't match the vacancy.